velox/exec/HashTable.cpp

```diff
@@ -613,22 +636,36 @@ void HashTable<ignoreNullKeys>::clear() {
 
 template <bool ignoreNullKeys>
 void HashTable<ignoreNullKeys>::checkSize(int32_t numNew) {
+  VELOX_CHECK(
+      size_ == 0 || size_ > (numDistinct_ + numTombstones_),
+      "size_:{} numDistinct_:{} numTombstoneRows_:{}",
+      size_,
+      numDistinct_,
+      numTombstones_);
+
+  const int64_t numDistincts = numNew + numDistinct_;
   if (!table_ || !size_) {
     // Initial guess of cardinality is double the first input batch or at
     // least 2K entries.
     // numDistinct_ is non-0 when switching from HashMode::kArray to regular
     // hashing.
     auto newSize = std::max(
         (uint64_t)2048, bits::nextPowerOfTwo(numNew * 2 + numDistinct_));
-    if (numNew + numDistinct_ > rehashSize(newSize)) {
+    if (numDistincts > rehashSize(newSize)) {
       newSize *= 2;
     }
     allocateTables(newSize);
-    if (numDistinct_) {
+    if (numDistinct_ > 0) {
       rehash();
     }
-  } else if (numNew + numDistinct_ > rehashSize()) {
-    auto newSize = bits::nextPowerOfTwo(size_ + numNew);
+    // We are not always able to reuse a tombstone slot as a free one for hash
+    // collision handling purpose. For example, if all the table slots are
+    // either occupied or tombstone, then we can't store any new entry in the
+    // table. Also, if there is non-trivial amount of tombstone slots in table,
+    // then the table lookup will become slow. Given that, we treat tombstone
+    // slot as non-empty slot here to decide whether to trigger rehash or not.
+  } else if (numDistincts > rehashSize(size_ - numTombstones_)) {
```

Comment:
Probably should change `rehashSize()` to return `rehashSize(size_ - numTombstones_)` so that `hashTableSizeIncrease` also has the right condition

Reply:
Good point!